lib/Sema/SemaDecl.cpp

```diff
@@ -11213,28 +11213,107 @@ void Sema::ActOnFinishKNRParamDeclarations(Scope *S, Declarator &D,
   }
 }
 
-// ActOnBoundsDecl: handle a Checked C bounds declaration for a declarator.
+static void checkBoundsDeclWithTypeAnnotation(Sema &S, DeclaratorDecl *D,
+                                             InteropTypeBoundsAnnotation *Expr,
+                                             bool IsReturnBounds) {
+  assert(D != nullptr && Expr != nullptr && !Expr->isInvalid());
+
+  QualType DeclaredTy = D->getType();
+  QualType AnnotTy = Expr->getType();
+  QualType OriginalAnnotTy = AnnotTy;
+
+  if (DeclaredTy.isNull() || AnnotTy.isNull())
+    return;
+
+  if (IsReturnBounds) {
+    assert(DeclaredTy->isFunctionType());
+    if (const FunctionType *FuncTy = DeclaredTy->getAs<FunctionType>())
+      DeclaredTy = FuncTy->getReturnType();
+    else
+      return;
+  }
+
+  if (isa<ParmVarDecl>(D)) {
+    AnnotTy = S.Context.getAdjustedParameterType(AnnotTy);
+  }
+
+  int DiagId = 0;
+  // Check that declared type is allowed to have a bounds-safe
+  // interface type declaration.
+  if (!DeclaredTy->isPointerType() && !DeclaredTy->isArrayType()) {
+    if (IsReturnBounds)
+      DiagId =
+        diag::err_typecheck_return_bounds_type_annotation_for_illegal_type;
+    else
+      DiagId = diag::err_typecheck_bounds_type_annotation_for_illegal_type;
+    S.Diag(Expr->getStartLoc(), DiagId);
+    return;
+  }
+
+   // Check that the annotation type is allowed to appear in a bound-safe
+  // interface type declaration.
+  SourceLocation AnnotTyLoc =
+    Expr->getTypeInfoAsWritten()->getTypeLoc().getBeginLoc();
+
+  if (IsReturnBounds) {
+    if (AnnotTy->isArrayType())
+      DiagId = diag::err_typecheck_return_bounds_type_annotation_is_array;
+    else if (!AnnotTy->isPointerType())
+      DiagId =
+        diag::err_typecheck_return_bounds_type_annotation_must_be_pointer;
+  } else {
+    if (!AnnotTy->isPointerType() && !AnnotTy->isArrayType())
+      DiagId = diag::err_typecheck_bounds_type_annotation_must_be_pointer_or_array;
+  }
+
+  if (DiagId) {
+    S.Diag(AnnotTyLoc, DiagId);
+    return;
+  }
+
+  // Check that the types are identical if checking is ignored.
+  if (!S.Context.isEqualIgnoringChecked(AnnotTy, DeclaredTy)) {
+    DiagId = diag::err_typecheck_bounds_type_annotation_incompatible;
+    S.Diag(AnnotTyLoc, DiagId) << AnnotTy << DeclaredTy;
+    return;
```

Comment:
The pattern in this function is to return as soon as we find an error.  Would it make more sense to generate all relevant errors before returning?  For example, if an annotated type is neither checked nor equivalent, should we report both issues at the same time?